doc/bulk_counter/bulk_counter.md

```diff
@@ -0,0 +1,145 @@
+  - SAI_STATS_MODE_READ -> SAI_STATS_MODE_BULK_READ
+  - SAI_STATS_MODE_READ_AND_CLEAR -> SAI_STATS_MODE_BULK_READ_AND_CLEAR
+
+### Architecture Design
+
+For each counter group, different statistic type is allowed to chooose bulk or non-bulk API based on SAI capability.
+
+![architecture](/doc/bulk_counter/bulk_counter.svg).
+
+> Note: In the picture, pg/queue watermark statistic use bulk API and buffer watermark statistic uses non-bulk API. This is just an example to show the design idea. 
+
+### High-Level Design
+
+Changes shall be made to sonic-sairedis to support this feature. No CLI change. No DB schema change.
+
+> Note: Code present in this design document is only for demonstrating the design idea, it is not production code.
+
+#### sonic-sairedis
+
+##### Bulk Statistic Context
+
+A new structure shall be added to FlexCounter class. 
+
+This structure is created because:
+
+- Meet the signature of sai_bulk_object_get_stats and sai_bulk_object_clear_stats
+- Avoid constructing these information each time collecting statistic. The bulk context shall only be updated under below cases:
+  - New object join counter group. E.g. adding a new port object.
+  - Existing object leave counter group. E.g removing an existing port object.
+
+```cpp
+struct BulkStatsContext
+{
+    bool support = true;
+    std::vector<sai_object_id_t> object_vids;
+    std::vector<sai_object_key_t> object_keys; 
+    std::vector<sai_stat_id_t> counter_ids;
+    std::vector<sai_status_t> get_statuses;
+    std::vector<sai_status_t> clear_statuses;
+    std::vector<uint64_t> counters;
+    std::shared_ptr<sai_stat_capability_list_t> stats_capas;
+};
+```
+- support: indicate if bulk is supported for this object type.
+- object_keys: objects that participate the bulk call. E.g. for port, SAI object id value shall be put into sai_object_key_t structure.
+- counter_ids: SAI statistic IDs that will be queried/cleared by the bulk call.
+- get_statuses: SAI bulk get return value for each object.
+- clear_statuses: SAI bulk clear return value for each object. Only effective if stats mode is read and clear.
+- counters: counter values that will be fill by vendor SAI.
+- stats_capas: stats capability for each statitstic IDs for current object type.
+
+The flow of how to updating bulk context will be discussed in following section.
+
+Instance of BulkStatsContext shall be added to FlexCounter class for each object type. One object type may have multiple BulkStatsContext instances.
+
+```cpp
+
+BulkStatsContext m_portBulkContext;
+BulkStatsContext m_priorityGroupBulkContext;
+BulkStatsContext m_queueBulkContext;
+...
+
+```
+
+##### Update Bulk Context
+
+1. New object join counter group.
+
+![Add Object Flow](/doc/bulk_counter/object_join_counter_group.svg).
+
+2. Existing object leave counter group, related data shall be removed from bulk context.
+
+##### Statistic Collect
+
+![Collect Counter Flow](/doc/bulk_counter/counter_collect.svg).
```

Comment:
1. If mode SAI_STATS_MODE_BULK_READ_AND_CLEAR is supported, call sai_bulk_object_get_stats is enough. I think it is equative to sai_bulk_object_get_stats(mode=BULK_READ) and then sai_bulk_object_clear_stats(mode=BULK_CLEAR).
2. If not sucess after sai_bulk_object_get_stats or sai_bulk_object_clear_stats, why go to 'Old way'?

Reply:
1. yes, you are right
2. we need keep backward compatible

Reply:
Shall we fall back to old way?